internal/registration/certs.go

```diff
@@ -0,0 +1,131 @@
+package registration
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rand"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+
+	"encoding/pem"
+
+	"git.sr.ht/~spc/go-log"
+	"github.com/project-flotta/flotta-operator/pkg/mtls"
+)
+
+type ClientCert struct {
+	keyPath   string
+	certPath  string
+	certGroup *mtls.CertificateGroup
+}
+
+func NewClientCert(certPath string, keyPath string) *ClientCert {
+	res := &ClientCert{
+		certPath: certPath,
+		keyPath:  keyPath,
+	}
+	return res
+}
+
+func (c *ClientCert) Init() error {
+	certPem, err := os.ReadFile(c.certPath)
+	if err != nil {
+		return err
+	}
+
+	certKey, err := os.ReadFile(c.keyPath)
+	if err != nil {
+		return err
+	}
+	c.certGroup = &mtls.CertificateGroup{
+		CertPEM:    bytes.NewBuffer(certPem),
+		PrivKeyPEM: bytes.NewBuffer(certKey),
+	}
+
+	err = c.certGroup.ImportFromPem()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (c *ClientCert) IsRegisterCert() (bool, error) {
+	if c.certGroup == nil {
+		return false, errors.New("Certificates are not imported")
+	}
+
+	cert := c.certGroup.GetCert()
+	if cert == nil {
+		return true, errors.New("No valid cert")
+	}
+
+	return cert.Subject.CommonName == mtls.CertRegisterCN, nil
+}
+
+func (c *ClientCert) Renew(deviceID string, key crypto.PrivateKey) ([]byte, []byte, error) {
+	var csrTemplate = x509.CertificateRequest{
+		Subject: pkix.Name{
+			CommonName: deviceID,
+			// Operator will add metadata on this subject, like namespace
+		},
+	}
+
+	csrCertificate, err := x509.CreateCertificateRequest(rand.Reader, &csrTemplate, key)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPEM := pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE REQUEST",
+		Bytes: csrCertificate,
+	})
+
+	keyPEM, err := c.certGroup.MarshalKeyToPem(key)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot encode new key: %v", err)
+	}
+
+	return certPEM, keyPEM.Bytes(), nil
+}
+
+func (c *ClientCert) CreateDeviceCerts(deviceID string) ([]byte, []byte, error) {
+	key, err := c.certGroup.GetNewKey()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	return c.Renew(deviceID, key)
+}
+
+func (c *ClientCert) WriteCertificate(cert, key []byte) error {
+	certGroup := &mtls.CertificateGroup{
+		CertPEM:    bytes.NewBuffer(cert),
+		PrivKeyPEM: bytes.NewBuffer(key),
+	}
+
+	err := certGroup.ImportFromPem()
+	if err != nil {
+		return fmt.Errorf("failing on importing certificate: %v", err)
+	}
+
+	err = ioutil.WriteFile(c.certPath, cert, 0600)
+	if err != nil {
+		return err
+	}
+
+	err = ioutil.WriteFile(c.keyPath, key, 0600)
```

Comment:
I think it should be an atomic functionality - 
If something went wrong here -with writing the keyPath you will have a new certPath but the keyPath will be the old one 

Maybe keep the old pair in a backup file

Reply:
If you look at the next lines, if key cannot be written it restore the old cert.

Reply:
I missed that it's using the certGroup as a backup